### PR TITLE
jewel: rbd: possible deadlock in various maintenance operations

### PR DESCRIPTION
**Pull request by smithfarm**
http://tracker.ceph.com/issues/22175

### CODE REVIEW

**Comment by smithfarm** [comment]
@dillaman @trociny This one was pretty non-trivial - if it's not easily salvageable, feel free to close it.

**Comment by smithfarm** [comment]
```
librbd/Operations.cc: In member function ‘int librbd::Operations<ImageCtxT>::snap_rollback(const char*, librbd::ProgressContext&)’:
librbd/Operations.cc:706:50: error: ‘snap_namespace’ was not declared in this scope
       uint64_t snap_id = m_image_ctx.get_snap_id(snap_namespace, snap_name);
                                                  ^
librbd/Operations.cc: In member function ‘int librbd::Operations<ImageCtxT>::prepare_image_update(bool)’:
librbd/Operations.cc:1159:7: error: redeclaration of ‘int r’
   int r = 0;
       ^
librbd/Operations.cc:1138:7: note: ‘int r’ previously declared here
   int r = 0;
       ^
librbd/Operations.cc: In instantiation of ‘int librbd::Operations<ImageCtxT>::prepare_image_update(bool) [with ImageCtxT = librbd::ImageCtx]’:
librbd/Operations.cc:1197:24:   required from here
librbd/Operations.cc:1152:9: error: ‘class librbd::ExclusiveLock<librbd::ImageCtx>’ has no member named ‘acquire_lock’
         m_image_ctx.exclusive_lock->acquire_lock(&ctx);
         ^
librbd/Operations.cc:1154:9: error: ‘class librbd::ExclusiveLock<librbd::ImageCtx>’ has no member named ‘try_acquire_lock’
         m_image_ctx.exclusive_lock->try_acquire_lock(&ctx);
         ^
Makefile:23016: recipe for target 'librbd/Operations.lo' failed
```

**Comment by smithfarm** [comment]
@dillaman @trociny This code block is problematic in jewel because methods "acquire_lock" and "try_acquire_lock" don't exist in the `exclusive_lock` class:

```
+      if (request_lock) {
+        m_image_ctx.exclusive_lock->acquire_lock(&ctx);
+      } else {
+        m_image_ctx.exclusive_lock->try_acquire_lock(&ctx);
+      }
```

**Comment by trociny** [comment]
@smithfarm You should use `request_lock` and `try_lock` methods instead.

File: src/librbd/Operations.cc [hunk 1]
```diff
@@ -1119,34 +1126,50 @@ void Operations<I>::execute_snap_unprotect(const std::string &snap_name,
...
 
   // need to upgrade to a write lock
   int r = 0;
-  bool trying_lock = false;
-  C_SaferCond ctx;
+
+  Context *ctx = util::create_async_context_callback(
+    m_image_ctx, util::create_context_callback<
+      C_InvokeAsyncRequest<I>,
+      &C_InvokeAsyncRequest<I>::handle_acquire_exclusive_lock>(this));
```
Comment (smithfarm) [inline src/librbd/Operations.cc]: @trociny How do I initialize `ctx` correctly here so I can send it to request_lock and try_lock, below?

Comment (trociny) [reply src/librbd/Operations.cc]: @smithfarm you don't need this. `C_SaferCond ctx` was perfectly good for this. What you only need is to replace:
```
+      if (request_lock) {
+        m_image_ctx.exclusive_lock->acquire_lock(&ctx);
+      } else {
+        m_image_ctx.exclusive_lock->try_acquire_lock(&ctx);
+      }
```
from the original patch with:
```
+      if (request_lock) {
+        m_image_ctx.exclusive_lock->request_lock(&ctx);
+      } else {
+        m_image_ctx.exclusive_lock->try_lock(&ctx);
+      }
```

**Comment by smithfarm** [comment]
```
[----------] 29 tests from TestMirroring
[ RUN      ] TestMirroring.EnableImageMirror_In_MirrorModeImage
./test/run-rbd-unit-tests.sh: line 9: 22683 Segmentation fault      (core dumped) unittest_librbd
FAIL test/run-rbd-unit-tests.sh (exit status: 139)
```

**Comment by smithfarm** [comment]
This passed a rados run at http://tracker.ceph.com/issues/21742#note-15

Still waiting for an rbd run.

**Comment by smithfarm** [comment]
@trociny @dillaman This PR was included in an rbd run at http://tracker.ceph.com/issues/21742#note-19

The results are not completely 100% convincing. The initial run had 3 failed jobs out of 154. On re-run, 2 of these succeeded, but one - rbd/qemu/{cache/writeback.yaml cachepool/none.yaml clusters/{fixed-3.yaml openstack.yaml} features/journaling.yaml msgr-failures/few.yaml objectstore/filestore-xfs.yaml workloads/qemu_xfstests.yaml} - failed a second time, and (looks like) for the same reason.

I'm running that one again, but I'd like to ask you to look at it please. The link above shows the rados and rbd PRs included in the run.

**Comment by trociny** [comment]
For these 3 failed jobs:
1) admin socket test failure. It failed on the start, on wait_for_clean, i.e. checking the current ceph status. It is not rbd related, probably a bug in wait_for_clean. Actually `wait_for_clean` does not look necessary in this scrip and I think we just have to remove it, as we did for journal test recently. I will create a PR for this.

2) qemu_xfstests failure. The test failed due to the osd crash:
```
2018-03-28T13:30:26.390 INFO:tasks.ceph.osd.3.smithi009.stderr:2018-03-28 13:30:26.388489 7f1076381700 -1 osd.3 pg_epoch: 8 pg[0.8( v 8'7164 (8'4100,8'7164] local-les=4 n=63 ec=1 les/c/f 4/4/0 3/3/3) [3,5] r=0 lpr=3 luod=8'7163 lua=8'7163 crt=8'7162 lcod 8'7162 mlcod 8'7162 active+clean] bad op order, already applied 56549 > this 56537
2018-03-28T13:30:26.390 INFO:tasks.ceph.osd.3.smithi009.stderr:osd/ReplicatedPG.cc: In function 'void ReplicatedPG::execute_ctx(ReplicatedPG::OpContext*)' thread 7f1076381700 time 2018-03-28 13:30:26.388498
2018-03-28T13:30:26.390 INFO:tasks.ceph.osd.3.smithi009.stderr:osd/ReplicatedPG.cc: 3044: FAILED assert(0 == "out of order op")
2018-03-28T13:30:26.390 INFO:tasks.ceph.osd.3.smithi009.stderr: ceph version 10.2.10-344-gff281b3 (ff281b3758cabe133a0a217330389069936be42e)
2018-03-28T13:30:26.390 INFO:tasks.ceph.osd.3.smithi009.stderr: 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x80) [0x5589c955c440]
2018-03-28T13:30:26.391 INFO:tasks.ceph.osd.3.smithi009.stderr: 2: (ReplicatedPG::execute_ctx(ReplicatedPG::OpContext*)+0x1990) [0x5589c8ffdf90]
2018-03-28T13:30:26.391 INFO:tasks.ceph.osd.3.smithi009.stderr: 3: (ReplicatedPG::do_op(std::shared_ptr<OpRequest>&)+0x3977) [0x5589c9001f37]
2018-03-28T13:30:26.391 INFO:tasks.ceph.osd.3.smithi009.stderr: 4: (ReplicatedPG::do_request(std::shared_ptr<OpRequest>&, ThreadPool::TPHandle&)+0x603) [0x5589c8fbda33]
2018-03-28T13:30:26.391 INFO:tasks.ceph.osd.3.smithi009.stderr: 5: (OSD::dequeue_op(boost::intrusive_ptr<PG>, std::shared_ptr<OpRequest>, ThreadPool::TPHandle&)+0x411) [0x5589c8e5ee91]
2018-03-28T13:30:26.391 INFO:tasks.ceph.osd.3.smithi009.stderr: 6: (PGQueueable::RunVis::operator()(std::shared_ptr<OpRequest>&)+0x52) [0x5589c8e5f0e2]
2018-03-28T13:30:26.391 INFO:tasks.ceph.osd.3.smithi009.stderr: 7: (OSD::ShardedOpWQ::_process(unsigned int, ceph::heartbeat_handle_d*)+0x6e1) [0x5589c8e7be21]
2018-03-28T13:30:26.391 INFO:tasks.ceph.osd.3.smithi009.stderr: 8: (ShardedThreadPool::shardedthreadpool_worker(unsigned int)+0x73a) [0x5589c954c41a]
2018-03-28T13:30:26.391 INFO:tasks.ceph.osd.3.smithi009.stderr: 9: (ShardedThreadPool::WorkThreadSharded::entry()+0x10) [0x5589c954f530]
2018-03-28T13:30:26.391 INFO:tasks.ceph.osd.3.smithi009.stderr: 10: (()+0x76ba) [0x7f1097b5b6ba]
2018-03-28T13:30:26.392 INFO:tasks.ceph.osd.3.smithi009.stderr: 11: (clone()+0x6d) [0x7f1095bd341d]
```
It is not rbd related, although it looks like an indication of a real problem in rados, and is probably worth investigating and reporting (I see several "out of order op" tickets in the tracker, but they all are in resolved state).

3) Object map rebuild test failure. This looks like a race in the test. It checks if the qemu is still running by listing the current image watchers. If the list is not empty it assumes that qemu is running and executes object map rebuild. But in this case the qemu just was stopped and rbd image removal started. And when the script checked for watchers it got non empty list due to `rbd remove` client, assumed it was qemu, tried object map rebuild and failed.
```
2018-03-28T14:16:37.026 INFO:tasks.qemu.client.0.smithi195.stdout:[ 2294.847787] Power down.
2018-03-28T14:16:37.344 DEBUG:tasks.qemu:checking that qemu tests succeeded...
2018-03-28T14:16:37.977 INFO:tasks.rbd:Deleting rbd clones...
2018-03-28T14:16:37.977 INFO:teuthology.orchestra.run.smithi195:Running: 'adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage rbd -p rbd rm client.0.1-clone'
2018-03-28T14:16:39.380 INFO:tasks.workunit.client.0.smithi195.stderr:+ is_qemu_running
2018-03-28T14:16:39.380 INFO:tasks.workunit.client.0.smithi195.stderr:+ rbd status client.0.1-clone
2018-03-28T14:16:39.381 INFO:tasks.workunit.client.0.smithi195.stderr:+ grep -v 'Watchers: none'
2018-03-28T14:16:39.442 INFO:tasks.workunit.client.0.smithi195.stdout:Watchers:
2018-03-28T14:16:39.442 INFO:tasks.workunit.client.0.smithi195.stdout:	watcher=172.21.15.195:0/935500 client.4236 cookie=139721862811120
2018-03-28T14:16:39.443 INFO:tasks.workunit.client.0.smithi195.stdout:*** Rebuilding object map
2018-03-28T14:16:39.443 INFO:tasks.workunit.client.0.smithi195.stderr:+ echo '*** Rebuilding object map'
2018-03-28T14:16:39.443 INFO:tasks.workunit.client.0.smithi195.stderr:+ rbd object-map rebuild client.0.1-clone
```

So I think the PR is ok to merge.

[1] http://qa-proxy.ceph.com/teuthology/smithfarm-2018-03-28_12:50:32-rbd-wip-jewel-backports-distro-basic-smithi/2330879/teuthology.log
[2] http://qa-proxy.ceph.com/teuthology/smithfarm-2018-03-28_12:50:32-rbd-wip-jewel-backports-distro-basic-smithi/2330958/teuthology.log
[3] http://qa-proxy.ceph.com/teuthology/smithfarm-2018-03-28_12:50:32-rbd-wip-jewel-backports-distro-basic-smithi/2330979/teuthology.log

**Comment by trociny** [comment]
> 3. Object map rebuild test failure. This looks like a race in the test.
I will create a tracker ticket for this too.

**Comment by smithfarm** [comment]
@trociny Thanks for analyzing the test failures.

**Comment by smithfarm** [comment]
@trociny I did a rados run on the same integration branch, and got the same "out of order op" OSD crashes there, too.

**Comment by smithfarm** [comment]
@trociny Ah, the "out of order op" assert is http://tracker.ceph.com/issues/18739 (won't fix for jewel)

**Comment by trociny** [comment]
@smithfarm Indeed.

I created the tracker tickets for two other test failures:
https://tracker.ceph.com/issues/23499
https://tracker.ceph.com/issues/23502

**Pull request merged**